lib/spack/spack/util/compression.py

```diff
@@ -39,6 +40,39 @@ def _gunzip(archive_file):
             f_out.write(f_in.read())
 
 
+def spack_unzipper():
+    from zipfile import ZipFile, ZipInfo
+
+    # implementation based on https://stackoverflow.com/a/39296577
+    class SpackZipFile(ZipFile):
+        if sys.version_info < (3, 6):
+            def extract(self, member, path=None, pwd=None):
+                # below is a direct copy of ZipFile's extract method
+                if not isinstance(member, ZipInfo):
+                    member = self.getinfo(member)
+                if path is None:
+                    path = os.getcwd()
+                ret_val = self._extract_member(member, path, pwd)
+
+                # catch output and restore permissions
+                os.chmod(ret_val, member.external_attr >> 16)
+                return ret_val
+
+        else:
+            def _extract_member(self, member, targetpath, pwd):
+                # Call ZipFile extract member method
+                if not isinstance(member, ZipInfo):
+                    member = self.getinfo(member)
+                path = super(ZipFile, self)._extract_member(member, targetpath, pwd)
+
+                # Catch output and restore permissions if permissions attr is non zero
+                if member.external_attr > 0xffff:
+                    os.chmod(path, member.external_attr >> 16)
```

Comment:
Normally I'd ask to have this documented, but according to https://bugs.python.org/issue18262 there is no documentation for this attribute (I do think that link would be useful in a comment here and possibly also https://bugs.python.org/issue15795, since that would cover the reason why there is fiddling with `external_attr` at all).

I'm therefore inclined to avoid using `zipfile` whenever possible (i.e. a hybrid of this and https://github.com/spack/spack/pull/30200 - do this stuff on Windows but don't use `zipfile` on other systems).

Reply:
I did not find that issue, but also struggled to find documentation, glad to know it wasn't because I was bad at searching. 

Is the hesitancy around using an undocumented variable? I'd rather avoid as much separation of the Windows/other platform interfaces as possible, and it seems like the issue would still exist on Windows as well, so IMO, we may as well develop an alternative solution that works for both platforms.
Creating separate interfaces for different platforms feels like an easy way to end up maintaining two separate codebases that exist on top of each other. Obviously there will be some unavoidable instances where we need to have truly separate interfaces (i.e. the `posixpath` vs `nt` modules) and this may be once of those scenarios, but I think putting some extra effort into working toward a unified solution would be worth it.
If you feel differently, I'll happily implement in the manner suggested in the above comment (and that may be what we need to do anyway if we can't determine a better method)

Reply:
> Is the hesitancy around using an undocumented variable?

Yes: if they don't document it, then they can change it whenever they want. Still, it is the only way to do things on Windows, so I think this is necessary. I think it's fragile though (for reasons we can't control) so we should only subject ourselves to that fragility when necessary.

Reply:
Ok that's kind of what I was thinking but wanted to be sure. I'm still hesitant to use something that fragile on Windows. We're only supporting Windows 10 and AFAIK Windows 10 comes with a tar executable that can unpack `.tar`, `.gz`, `.zip` files. Additionally Powershell has a method to do this, or we can write some visual basic wrapped in a `.bat` file that Spack can invoke via `Executable`, so if it's alright with you, I'd prefer to refactor to avoiding `ZipFile` altogether and to explore one of these alternatives.

Reply:
> if it's alright with you, I'd prefer to refactor to avoid ZipFile altogether and to explore one of these alternatives.

That sounds good to me